File: 02_Object_Oriented_Programming/02_Class_Relationships/02_Association/javascript/person.js
/**
 * Association — Person class in JavaScript.
 *
 * A Person is associated with a Car.
 * The Car is passed in from outside — Person does NOT create it.
 */

class Person {
  #name;
  #car; // Association: Person "has a" Car

  constructor(name, car) {
    this.#name = name;
    this.#car = car;
  }

  get name() {
    return this.#name;
  }

  get car() {
    return this.#car;
  }

  set car(car) {
    this.#car = car;
  }

  goForDrive() {
    console.log(`${this.#name} is going for a drive.`);
    this.#car.drive();
  }

  toString() {
    return `Person(${this.#name}, car=${this.#car})`;
  }
}

module.exports = Person;

File: 02_Object_Oriented_Programming/02_Class_Relationships/02_Association/java/Person.java
/**
 * Association — Person class.
 *
 * A Person is associated with a Car.
 * The Car is passed in from outside — Person does NOT create it.
 * Both can exist independently.
 */
public class Person {

    private String name;
    private Car car;  // Association: Person "has a" Car

    public Person(String name, Car car) {
        this.name = name;
        this.car = car;
    }

    public String getName() {
        return name;
    }

    public Car getCar() {
        return car;
    }

    public void setCar(Car car) {
        this.car = car;
    }

    public void goForDrive() {
        System.out.println(name + " is going for a drive.");
        car.drive();
    }

    @Override
    public String toString() {
        return "Person(" + name + ", car=" + car + ")";
    }
}
